hera_cal/io.py

```diff
@@ -864,6 +864,147 @@ def iterate_over_times(self, Nints=1, times=None):
+                               for n, (i, j) in enumerate(zip(ant1_array,
+                                                              ant2_array))}
+                if 'bls' not in info:
+                    info['bls'] = set(inds[_hash].keys())
+                else:
+                    info['bls'].intersection_update(set(inds[_hash].keys()))
+            bl2ind[filename] = inds[_hash]
+
+    if bls is None:
+        # generate a set of bls if we didn't have one passed in
+        if pols is None:
+            pols = list(pol_indices.keys())
+        bls = info['bls']
+        bls = set(bl + (p,) for bl in bls for p in pols)
+    else:
+        # if length 2 baselines are passed in, add on polarizations
+        bls_len2 = set(bl for bl in bls if len(bl) == 2)
+        if len(bls_len2) > 0:
+            if pols is None:
+                pols = list(pol_indices.keys())
+            bls = set(bl for bl in bls if len(bl) == 3)
+            bls = bls.union([bl + (p,) for bl in bls_len2 for p in pols])
+        # record polarizations as total of ones indexed in bls
+        pols = set(bl[2] for bl in bls)
+    # sort files by time of first integration
+    times.sort(key=lambda x: x[0][0])
+    filenames = (v[-1] for v in times)
+    times = np.concatenate([t[0] for t in times], axis=0)
+    info['times'] = times
+
+    # preallocate buffers
+    rv = {}
+    if read_data:
+        rv['visdata'] = {bl: np.empty((times.size, info['freqs'].size), dtype=dtype) for bl in bls}
+    if read_flags:
+        rv['flags'] = {bl: np.empty((times.size, info['freqs'].size), dtype=bool) for bl in bls}
+    if read_nsamples:
+        rv['nsamples'] = {bl: np.empty((times.size, info['freqs'].size), dtype=np.float32) for bl in bls}
+    # bail here if all we wanted was the info
+    if len(rv) == 0:
+        return {'info': info}
+
+    t = 0
+    for filename in filenames:
+        inds = bl2ind[filename]
+        if verbose:
+            print(f'Reading data from {filename}')
+        with h5py.File(filename, 'r') as f:
+            h = f['/Header']
+            ntimes = int(h['Ntimes'][()])
+            nbls = int(h['Nblts'][()]) // ntimes
+            if check:
+                assert np.allclose(h['time_array'][:ntimes], times[t:t + ntimes])
+            # decide whether to read all the data in, or use partial I/O
+            full_read = (len(bls) > full_read_thresh * nbls * npols)
+            if full_read and verbose:
+                print('Reading full file')
+            for key, data in rv.items():
+                d = f['/Data'][key]  # data not read yet
+                if full_read:
+                    d = d[()]  # reads data
+                # handle HERA's raw (int) and calibrated (complex) file formats
+                if key == 'visdata' and not np.iscomplexobj(d):
+                    # Support polarization-transposed arrays
+                    if d.shape[-1] == info['freqs'].size:
```

Comment:
I think you can save a few lines of code if you put this if inside the for loop

Reply:
I purposely unrolled to not do if checks inside the for loop, which is slower